pkg/controller/nodes/task/k8s/event_watcher.go

```diff
@@ -0,0 +1,119 @@
+	mu         sync.RWMutex
+	eventInfos map[types.NamespacedName]*EventInfo
+}
+
+// EventInfo stores detail about the event and the timestamp of the first occurrence.
+// All other fields are thrown away to conserve space.
+type EventInfo struct {
+	Reason     string
+	Note       string
+	CreatedAt  time.Time
+	RecordedAt time.Time
+}
+
+func (e *eventWatcher) OnAdd(obj interface{}) {
+	event := obj.(*eventsv1.Event)
+	objectNsName := types.NamespacedName{Namespace: event.Regarding.Namespace, Name: event.Regarding.Name}
+	eventNsName := types.NamespacedName{Namespace: event.Namespace, Name: event.Name}
+	v, _ := e.objectCache.LoadOrStore(objectNsName, &objectEvents{
+		eventInfos: map[types.NamespacedName]*EventInfo{},
+	})
+	objEvents := v.(*objectEvents)
+	objEvents.mu.Lock()
+	defer objEvents.mu.Unlock()
+	objEvents.eventInfos[eventNsName] = &EventInfo{
+		Reason:     event.Reason,
+		Note:       event.Note,
+		CreatedAt:  event.CreationTimestamp.Time,
+		RecordedAt: time.Now(),
+	}
+}
+
+func (e *eventWatcher) OnUpdate(_, newObj interface{}) {
+	// Dropping event updates since we only care about the creation
+}
+
+func (e *eventWatcher) OnDelete(obj interface{}) {
+	event := obj.(*eventsv1.Event)
+	objectNsName := types.NamespacedName{Namespace: event.Regarding.Namespace, Name: event.Regarding.Name}
+	eventNsName := types.NamespacedName{Namespace: event.Namespace, Name: event.Name}
+	v, _ := e.objectCache.LoadOrStore(objectNsName, &objectEvents{})
+	objEvents := v.(*objectEvents)
+	objEvents.mu.Lock()
+	defer objEvents.mu.Unlock()
+	delete(objEvents.eventInfos, eventNsName)
+	if len(objEvents.eventInfos) == 0 {
+		e.objectCache.Delete(objectNsName)
+	}
+}
+
+// List returns all events for the given object that were created after the given time, sorted by creation time.
+func (e *eventWatcher) List(objectNsName types.NamespacedName, createdAfter time.Time) []*EventInfo {
+	v, _ := e.objectCache.Load(objectNsName)
+	if v == nil {
+		return []*EventInfo{}
+	}
+	objEvents := v.(*objectEvents)
+	objEvents.mu.RLock()
+	defer objEvents.mu.RUnlock()
+	// This logic assumes that cardinality of events per object is relatively low, so iterating over them to find
+	// recent ones and sorting the results is not too expensive.
+	result := make([]*EventInfo, 0, len(objEvents.eventInfos))
+	for _, eventInfo := range objEvents.eventInfos {
+		if eventInfo.CreatedAt.After(createdAfter) {
+			result = append(result, eventInfo)
+		}
+	}
+	sort.SliceStable(result, func(i, j int) bool {
+		return result[i].CreatedAt.Before(result[j].CreatedAt) ||
+			(result[i].CreatedAt.Equal(result[j].CreatedAt) && result[i].RecordedAt.Before(result[j].RecordedAt))
+	})
+	return result
+}
+
+func NewEventWatcher(ctx context.Context, gvk schema.GroupVersionKind, kubeClientset kubernetes.Interface) (EventWatcher, error) {
+	objectSelector := func(opts *metav1.ListOptions) {
+		opts.FieldSelector = fields.OneTermEqualSelector("regarding.kind", gvk.Kind).String()
```

Comment:
sorry if I'm missing something here but why not use pluginEntry.ResourceToWatch here?